relay-general/src/store/transactions/processor.rs

```diff
@@ -53,28 +50,20 @@ impl<'r> TransactionsProcessor<'r> {
                 transaction.push('/');
             }
 
-            // Returns true if the source of the transaction matches provided source.
-            let source_match = |source: &TransactionSource| {
+            let rule = self.tx_name_rules.iter().find(|rule| {
                 info.as_ref()
-                    .and_then(|i| i.source.value().map(|s| s == source))
+                    .map(|trans_info| rule.matches(transaction, trans_info))
                     .unwrap_or_default()
-            };
-
-            let rule = self.tx_name_rules.iter().find(|rule| {
-                source_match(&rule.scope.source)
-                    && rule.expiry > now
-                    // Adding `/` at the end of the name, ensures that rules like /<something>/*/**
-                    // will always match the string.
-                    && rule.pattern.is_match(transaction)
             });
             let result = rule.map(|r| (r.pattern.pattern(), r.apply(transaction)));
 
-            if let Some((rule, Some(mut result))) = result {
-                if !slash_is_present {
-                    transaction.pop();
-                    result.pop();
-                }
+            if let Some((rule, mut result)) = result {
                 if &result != transaction {
+                    if !slash_is_present {
+                        transaction.pop();
```

Comment:
I think we should copy these two lines to line 59, that is, re-remove the added slash even if no rule matches at all. In other words, the original, unmodified transaction name should always survive. Maybe it's safer to use a `Cow` for the transaction name,  and evaluate the rule matching / replacement on a copy when `slash_is_present`.

Reply:
Currently the slash is always removed if it was added, on the line `77` is the check, even if the rule is not matching. 

I was thinking to copy the string and use it for the checks and manipulations - but I also want to avoid unnecessary copies if we can avoid it.

But I see that the code now a bit longer, and keeping code more compact would help here. I'll look into this.

Reply:
Sorry, I missed that. Yeah, to make this logic more resilient against future refactors it would make sense to make it more compact. Either by using a `Cow` or by something like a context manager as @iker-barriocanal suggested. The rust equivalent to a context manager would be the [RAII](https://doc.rust-lang.org/rust-by-example/scope/raii.html) pattern.

Reply:
@jjbayer  have a look into this commit [8199c5c](https://github.com/getsentry/relay/pull/1695/commits/8199c5caebfcf27b5d8fb11c1650bd8a4338334a), I tried to use `Cow` and hide the checks in the rule impl:
* it doesn't change the transaction
* it uses owned data only if needed
* and also combined the check and apply into one function